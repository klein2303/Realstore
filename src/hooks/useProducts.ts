import {useQuery, UseQueryResult} from "@tanstack/react-query";
import {Product} from "../models/Product";

const fetchProducts = async (
    category?: string,
    productId?: number,
    minPrice?: number,
    maxPrice?: number,
    minRating?: number,
): Promise<Product[]> => {
    let url = "https://fakestoreapi.com/products";

    if (productId) {
        url = `${url}/${productId}`;
        const response = await fetch(url);
        if (!response.ok) throw new Error("Error fetching product");
        const product: Product = await response.json();
        return [product]; // Return a single product wrapped in an array
    } else if (category) {
        url = `${url}/category/${category}`;
    }

    const response = await fetch(url);
    if (!response.ok) throw new Error("Error fetching products");
    let products: Product[] = await response.json();

    // Filter products based on price and rating
    if (minPrice !== undefined) {
        products = products.filter((product) => product.price >= minPrice);
    }
    if (maxPrice !== undefined) {
        products = products.filter((product) => product.price <= maxPrice);
    }
    if (minRating !== undefined) {
        products = products.filter((product) => product.rating.rate >= minRating);
    }

    return products;
};

// Custom hook to fetch products with optional filters
export const useProducts = (
    category?: string,
    productId?: number,
    minPrice?: number,
    maxPrice?: number,
    minRating?: number,
): UseQueryResult<Product[]> => {
    const queryKey = ["products", category, productId, minPrice, maxPrice, minRating];

    return useQuery<Product[]>({
        queryKey,
        queryFn: () => fetchProducts(category, productId, minPrice, maxPrice, minRating),
    });
};
