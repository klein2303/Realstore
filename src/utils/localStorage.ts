// Function to store favorite items in local storage
export const setFavorites = (items: number[]): void => {
    localStorage.setItem("favorites", JSON.stringify(items));
};

// Function to get favorite items from local storage
export const getFavorites = (): number[] => {
    const storedItems = localStorage.getItem("favorites");
    return storedItems ? JSON.parse(storedItems) : [];
};

export const toggleFavorite = (productId: number): void => {
    const favorites = getFavorites();
    const isFavorite = favorites.includes(productId);

    const updatedFavorites = isFavorite ? favorites.filter((id) => id !== productId) : [...favorites, productId];

    setFavorites(updatedFavorites);
};
