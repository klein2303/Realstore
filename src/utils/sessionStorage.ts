import {Filter} from "../pages/Shopping";

export type CartItem = {productId: number; quantity: number};

// Helper function to store items in session storage
const setSessionStorageItems = <T>(key: string, items: T[]): void => {
    sessionStorage.setItem(key, JSON.stringify(items));
};

// Helper function to get items from session storage
const getSessionStorageItems = <T>(key: string): T[] => {
    const storedItems = sessionStorage.getItem(key);
    return storedItems ? JSON.parse(storedItems) : [];
};

// Cart-specific functions
export const getCart = (): CartItem[] => {
    return getSessionStorageItems<CartItem>("cart");
};

export const addToCart = (productId: number, quantity: number = 1): void => {
    const cart = getCart();
    const existingItem = cart.find((item) => item.productId === productId);

    if (existingItem) {
        existingItem.quantity += quantity; // Increment quantity
    } else {
        cart.push({productId, quantity});
    }

    setSessionStorageItems("cart", cart);
};

export const clearCart = (): void => {
    setSessionStorageItems("cart", []);
};

export const updateCartQuantity = (productId: number, quantity: number): void => {
    const cart = getCart();
    const index = cart.findIndex((item) => item.productId === productId);

    if (index !== -1) {
        // Product exists in the cart
        if (quantity > 0) {
            cart[index].quantity = quantity; // Update quantity
        } else {
            cart.splice(index, 1); // Remove if quantity is 0
        }
    } else if (quantity > 0) {
        // Product doesn't exist in the cart, add it
        cart.push({productId, quantity});
    }

    setSessionStorageItems("cart", cart); // Save updated cart to sessionStorage
};

// Filtered items-specific functions
export const getFilter = (): Filter[] => {
    return getSessionStorageItems<Filter>("filter");
};

export const setFilter = (filter: Filter[]): void => {
    setSessionStorageItems("filter", filter);
};

export const clearFilter = (): void => {
    setSessionStorageItems("filter", []);
};
