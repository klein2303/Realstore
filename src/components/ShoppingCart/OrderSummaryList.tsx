import {useEffect, useState} from "react";
import styles from "./OrderSummaryList.module.css";
import {useProducts} from "../../hooks/useProducts";
import {Product} from "../../models/Product";
import {getCart, clearCart, CartItem} from "../../utils/sessionStorage";
import {useNavigate} from "react-router-dom";

const OrderSummaryList = () => {
    const [clicked, setClicked] = useState(false);
    const [price, setPrice] = useState<number>();
    const [cartItemData, setCartItemData] = useState<CartItem[]>([]);
    const {data: products, isLoading, isError} = useProducts();
    const navigate = useNavigate();

    useEffect(() => {
        if (!products) {
            return;
        }
        const cartItemData = getCart();
        setCartItemData(cartItemData);
        const cartItemIds = cartItemData.map((cartId: CartItem) => {
            return cartId.productId;
        });
        const currentPrice = products
            .filter((product: Product) => cartItemIds.includes(product.id))
            .map((product: Product) => product.price)
            .reduce((x: number, y: number) => (x = x + y), 0);
        setPrice(currentPrice);
    }, [products]);

    if (isLoading) {
        return (
            <section role="status" aria-live="polite" aria-label="Loading">
                Loading...
            </section>
        );
    }
    if (isError) {
        return (
            <section role="alert" aria-live="assertive" aria-label="Error fetching products">
                Error fetching products.
            </section>
        );
    }

    function handleChange(e: React.ChangeEvent<HTMLInputElement>) {
        const isChecked = e.target.checked;
        setClicked(isChecked);
    }

    function handleClick() {
        if (cartItemData) {
            clearCart();
            navigate(0);
        }
    }

    return (
        <>
            <section className={styles.box} role="region" aria-label="Order Summary Section">
                <h2 className={styles.title}>ORDER SUMMARY</h2>
                <div className={styles.textbox} aria-label="Subtotal Information">
                    <p className={styles.text}>Subtotal</p>
                    <p className={styles.text}>$ {price}</p>
                </div>
                <div className={styles.textbox} aria-label="Shipping Information">
                    <p className={styles.text}>Shipping</p>
                    <p className={styles.text}>$ Free</p>
                </div>
                <div className={styles.totalbox} aria-label="Total Cost">
                    <p className={styles.totaltext}>TOTAL</p>
                    <p className={styles.taxtext}>(TAX INCL.)</p>
                    <p className={styles.totaltext}>$ {price}</p>
                </div>
                <label className={styles.checkbox}>
                    <input
                        type="checkbox"
                        className={styles.check}
                        tabIndex={0}
                        onChange={(e) => handleChange(e)}
                        onKeyDown={(e) => {
                            if (e.key === "Enter") {
                                e.preventDefault(); // Prevent default scrolling or form submission behavior
                                const target = e.target as HTMLInputElement;
                                target.checked = !target.checked;
                                handleChange({
                                    ...e,
                                    target,
                                } as React.ChangeEvent<HTMLInputElement>);
                            }
                        }}
                        aria-labelledby="termsCheckbox"
                    />
                    <p id="termsCheckbox" className={styles.checkboxtext}>
                        I agree to the Terms and Conditions
                    </p>
                </label>

                <button
                    type="button"
                    disabled={!clicked}
                    aria-disabled={!clicked}
                    className={styles.button}
                    onClick={() => handleClick()}>
                    PAY
                </button>
            </section>
        </>
    );
};

export default OrderSummaryList;
