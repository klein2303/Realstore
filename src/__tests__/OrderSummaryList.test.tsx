import {render, screen, fireEvent} from "@testing-library/react";
import '@testing-library/jest-dom';
import {describe, it, expect, vi, beforeEach} from "vitest";
import OrderSummaryList from "../components/ShoppingCart/OrderSummaryList";
import {useProducts} from "../hooks/useProducts";
import {getCart, clearCart} from "../utils/sessionStorage";
import {MemoryRouter, useNavigate} from "react-router-dom";

vi.mock("../hooks/useProducts", () => ({
    useProducts: vi.fn(),
}));

vi.mock("../utils/sessionStorage", () => ({
    getCart: vi.fn(),
    clearCart: vi.fn(),
}));

vi.mock("react-router-dom", async () => {
    const actual = await vi.importActual("react-router-dom");
    return {
        ...actual,
        useNavigate: vi.fn(),
    };
});

describe("OrderSummaryList Component", () => {
    const mockNavigate = vi.fn();

    beforeEach(() => {
        vi.resetAllMocks();
        (useNavigate as jest.Mock).mockReturnValue(mockNavigate);
    });

    it("renders the loading state", () => {
        (useProducts as jest.Mock).mockReturnValue({
            data: null,
            isLoading: true,
            isError: false,
        });

        render(
            <MemoryRouter>
                <OrderSummaryList />
            </MemoryRouter>,
        );

        expect(screen.getByRole("status", {name: /loading/i})).toBeInTheDocument();
    });

    it("renders the error state", () => {
        (useProducts as jest.Mock).mockReturnValue({
            data: null,
            isLoading: false,
            isError: true,
        });

        render(
            <MemoryRouter>
                <OrderSummaryList />
            </MemoryRouter>,
        );

        expect(screen.getByRole("alert", {name: /error fetching products/i})).toBeInTheDocument();
    });

    it("renders the order summary with calculated price", () => {
        const mockProducts = [
            {id: 1, title: "Product 1", price: 100},
            {id: 2, title: "Product 2", price: 200},
        ];

        (useProducts as jest.Mock).mockReturnValue({
            data: mockProducts,
            isLoading: false,
            isError: false,
        });

        (getCart as jest.Mock).mockReturnValue([{productId: 1}, {productId: 2}]);

        render(
            <MemoryRouter>
                <OrderSummaryList />
            </MemoryRouter>,
        );

        expect(screen.getByLabelText(/subtotal information/i)).toHaveTextContent("$ 300");
        expect(screen.getByLabelText(/total cost/i)).toHaveTextContent("$ 300");
    });

    it("disables the PAY button until terms are accepted", () => {
        const mockProducts = [{id: 1, title: "Product 1", price: 100}];

        (useProducts as jest.Mock).mockReturnValue({
            data: mockProducts,
            isLoading: false,
            isError: false,
        });

        (getCart as jest.Mock).mockReturnValue([{productId: 1}]);

        render(
            <MemoryRouter>
                <OrderSummaryList />
            </MemoryRouter>,
        );

        const payButton = screen.getByRole("button", {name: /pay/i});
        expect(payButton).toBeDisabled();

        const termsCheckbox = screen.getByRole("checkbox", {name: /terms and conditions/i});
        fireEvent.click(termsCheckbox);

        expect(payButton).toBeEnabled();
    });

    it("clears the cart and refreshes the page when PAY is clicked", () => {
        const mockProducts = [{id: 1, title: "Product 1", price: 100}];

        (useProducts as jest.Mock).mockReturnValue({
            data: mockProducts,
            isLoading: false,
            isError: false,
        });

        (getCart as jest.Mock).mockReturnValue([{productId: 1}]);

        render(
            <MemoryRouter>
                <OrderSummaryList />
            </MemoryRouter>,
        );

        const termsCheckbox = screen.getByRole("checkbox", {name: /terms and conditions/i});
        const payButton = screen.getByRole("button", {name: /pay/i});

        fireEvent.click(termsCheckbox);
        expect(payButton).toBeEnabled();

        fireEvent.click(payButton);

        expect(clearCart).toHaveBeenCalled();
        expect(mockNavigate).toHaveBeenCalledWith(0);
    });

    it("matches the snapshot", () => {
        const mockProducts = [{id: 1, title: "Product 1", price: 100}];

        (useProducts as jest.Mock).mockReturnValue({
            data: mockProducts,
            isLoading: false,
            isError: false,
        });

        (getCart as jest.Mock).mockReturnValue([{productId: 1}]);

        const {container} = render(
            <MemoryRouter>
                <OrderSummaryList />
            </MemoryRouter>,
        );

        expect(container).toMatchSnapshot();
    });
});
