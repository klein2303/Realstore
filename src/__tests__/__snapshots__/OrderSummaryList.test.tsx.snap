// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OrderSummaryList Component > matches the snapshot 1`] = `
<div>
  <section
    aria-label="Order Summary Section"
    class="_box_3cb8da"
    role="region"
  >
    <h2
      class="_title_3cb8da"
    >
      ORDER SUMMARY
    </h2>
    <div
      aria-label="Subtotal Information"
      class="_textbox_3cb8da"
    >
      <p
        class="_text_3cb8da"
      >
        Subtotal
      </p>
      <p
        class="_text_3cb8da"
      >
        $ 
        100
      </p>
    </div>
    <div
      aria-label="Shipping Information"
      class="_textbox_3cb8da"
    >
      <p
        class="_text_3cb8da"
      >
        Shipping
      </p>
      <p
        class="_text_3cb8da"
      >
        $ Free
      </p>
    </div>
    <div
      aria-label="Total Cost"
      class="_totalbox_3cb8da"
    >
      <p
        class="_totaltext_3cb8da"
      >
        TOTAL
      </p>
      <p
        class="_taxtext_3cb8da"
      >
        (TAX INCL.)
      </p>
      <p
        class="_totaltext_3cb8da"
      >
        $ 
        100
      </p>
    </div>
    <label
      class="_checkbox_3cb8da"
    >
      <input
        aria-labelledby="termsCheckbox"
        class="_check_3cb8da"
        tabindex="0"
        type="checkbox"
      />
      <p
        class="_checkboxtext_3cb8da"
        id="termsCheckbox"
      >
        I agree to the Terms and Conditions
      </p>
    </label>
    <button
      aria-disabled="true"
      class="_button_3cb8da"
      disabled=""
      type="button"
    >
      PAY
    </button>
  </section>
</div>
`;
